device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -530,18 +530,18 @@ private void addProxyLayer(Transport transport, String hostName)
     private void sendQueuedMessages()
     {
         int messagesAttemptedToBeProcessed = 0;
-        boolean lastSendSucceeded = true;
         Message message = messagesToSend.poll();
-        while (message != null && messagesAttemptedToBeProcessed < MAX_MESSAGES_TO_SEND_PER_CALLBACK && lastSendSucceeded)
+        while (message != null && messagesAttemptedToBeProcessed < MAX_MESSAGES_TO_SEND_PER_CALLBACK)
         {
             messagesAttemptedToBeProcessed++;
-            lastSendSucceeded = sendQueuedMessage(message);
+            boolean lastSendSucceeded = sendQueuedMessage(message);
 
             if (!lastSendSucceeded)
             {
                 //message failed to send, likely due to lack of link credit available. Re-queue and try again later
                 log.trace("Amqp message failed to send, adding it back to messages to send queue ({})", message);
                 messagesToSend.add(message);
+                return;
```

Comment:
It's important to return here before the next call to messagesToSend.poll() as it keeps the next message to send at the front of the queue. Before this change, if exactly two messages are queued, and the first fails to send (if twin links aren't up yet, for instance), then the second message will never send.

Reply:
Did you verify only with the e2e test not failing or something else?

Reply:
I verified by manually causing that 2 message scenario and verifying that the SDK eventually sends the subscription message. There isn't a way to test this reliably because we don't have control over when each message will be queued from the e2e tests level. There is some level of randomness